william-wtr92/instamint
# feat(login): 🚀 users login / cookies & session management

# As a MINTER, I can log in to my account, when I enter my e-mail and password [ITM-16](https://project-william.atlassian.net/browse/ITM-16)

This task required users to be able to log in, and we took the safest possible decision by managing sessions with the creation of signed cookies in the backend.
Users can login from a fully translated and responsive interface.

## Proposed Changes

The emphasis has really been on making the login easy to use, with specific, translated error handling.

- Added : 
  - View: Login Form with errors management dynamically and fully translated
  - API route for Login 
  - Chore: Caching of active JWT token in Redis & delete of the previous one
  - Chore: Signed Secure Cookies implementation in Headers when users login
  - Chore: Implementation of SWR Config & Context
  - View: Get User Connected with SWR Hook
  - Chore: Protection of routes requiring login with NextJS `middleware` functionality 
  - Chore: Customization of CORS with credential and origin
  - Chore: Avatar Component 
  - Documentation: 
     - SWR
     - Middleware NextJS
     - Cookies in HonoJS 
  - View: Add redirection below login and register 
  - Chore: Actions Context for re-used actions like error &  success state / redirection delay & link
  - Chore: Unification of responses containing API data in a `result` object.
  - Chore: Hono updated to `4.1.7`
  - Chore: Constants for status codes
   
- Refactored:
   - Chore: Big refactor of ressources and path naming : `/users/..` to `/auth/...` 
   - Chore: Refactor of scafolding
   - Chore: Enhancement of type-safety
   - Chore: Factorisation of the JWT methods used several times on our bakend

- Updated:
   - Chore: Docker Compose updated with new env vars
   - Chore: `.env.example` updated 

- Fixed:
  - `post` method on email validation route change to `put`   
  - `tailwind.config` fixed `content` properties
  - `scafolding` in json `i18n` files

## Type of Change

What type of change are you making? (Check all that apply)


## How Has This Been Tested?

Explain the tests you conducted to validate your changes. Provide clear instructions so that the code review can
replicate your tests and confirm the results.

- To test this ticket, follow these instructions:
  - Run at root `pnpm install`
  - Update your `env variables` to `.env.example`
  - Run the api and the front end, then test the functionality on the following paths:
      - http://localhost:3000/auth/sign-in : Once you have created a user, go to the login page and log in. 
      - http://localhost:3000/ : Try accessing the root path when you are logged out and you will be redirected to /auth/sign-in

## Checklist

Before submitting your PR, please check the appropriate boxes to confirm that you have taken the following actions:


## Additional Information

- You can test with Docker  with `make dev-no-cache` & `make dev-up` commands.
- Don't hesitate to read the documentation provided in /docs in depth, which will help you understand the choice of tools and the architecture.

## Screenshots
#### Auth workflow schema of Instamint

<img width="1610" alt="Capture d’écran 2024-03-30 à 19 29 04" src="https://github.com/william-wtr92/instamint/assets/118012461/83d2b881-ccd9-4294-b6ef-311b67905542">




[ITM-16]: https://project-william.atlassian.net/browse/ITM-16?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
